Add multi-field search `search_fields` parameter to products REST API

staskus opened the pull request:
Original PR: https://github.com/woocommerce/woocommerce/pull/59241
RFC: pfZP8i-cI-p2 -worth checking it first
WOOMOB-713
Documentation PR: https://github.com/woocommerce/woocommerce-rest-api-docs/pull/271

## Changes proposed in this Pull Request:

This pull request introduces a new, flexible search mechanism to the WooCommerce Products REST API (`GET /wc/v3/products`). Instead of expanding `search_name_or_sku` parameter even further to support `global_unique_id`, this implementation uses a combination of the existing `search` parameter and a new `search_fields` array parameter. This allows clients to specify which fields to search (`name`, `sku`, `global_unique_id`, `description`, `short_description`) and supports cross-field matching logic.

The search behavior has been updated to support cross-field matching same way as a `search=` parameter:

* **Tokenized**: The `search` argument is treated as a collection of space-separated tokens.
* **Cross-Field Matching**: All tokens must be present across any of the specified fields for a product to match. A product matches if all tokens are found distributed across the selected fields.
* **Partial Match**: A word containing a token is a valid match.

**Examples:**

- An argument value of `blue shirt` with `search_fields=name` will match a product with the name "Casual Blue Shirt" because both tokens exist in the same field (name), but not "Red Shirt".
- `scarf 987` with `search_fields=name,global_unique_id` will match a product with name "Premium Wool Scarf" and GTIN "9876543210" because the tokens can be found across the specified fields.
- `quality professionals` with `search_fields=description,short_description` will match a product where "quality" appears in the short description and "professionals" appears in the description.

This new approach takes precedence over all other search-related parameters: `search_name_or_sku`, `search_sku`, and `sku`. If `search_fields` is present in a request, all others will be ignored for search logic.

This functionality is added to the v3 REST API products endpoint only.

## How to test the changes in this Pull Request:

1.  Create products with the following details:
    *   Product A:
        *   Name: `Premium Wool Scarf`
        *   SKU: `SCARF-W-PREM`
        *   GTIN: `9876543210`
        *   Description: `Essential tool for professionals.`
        *   Short Description: `High quality craftsmanship.`
    *   Product B:
        *   Name: `Casual Blue Shirt`
        *   SKU: `SHIRT-B-CASUAL`
        *   GTIN: `1234567890`
        *   Description: `A premium quality winter scarf made from wool.`
        *   Short Description: `Perfect for summer activities.`
    *   Product C:
        *   Name: `Red Cotton Tee`
        *   SKU: `TEE-C-RED`
        *   GTIN: `0000000000`

2.  Create an application password for your WordPress user in your user profile page from the WP admin area.

3.  **Test same-field matching:** Perform the following request to find "Product B" by searching for tokens that exist in the same field (name).

    ```bash
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=name&search=blue+shirt"
    ```

    You should get one result: the "Casual Blue Shirt" product.

4.  **Test individual field searches:** Verify that single tokens work across all fields.

    ```bash
    # Find by name token
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=name&search=scarf"

    # Find by SKU token  
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=sku&search=SCARF"

    # Find by GTIN token
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=global_unique_id&search=987"

    # Find by description token
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=description&search=winter"

    # Find by short description token
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=short_description&search=summer"
    ```

5.  **Test cross-field matching:** Confirm that tokens can match across different specified fields.

    ```bash
    # This should return results (tokens match across different fields)
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=name,global_unique_id&search=scarf+987"

    # Test cross-field matching with content fields
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=description,short_description&search=quality+professionals"
    ```

6.  **Verify precedence:** Confirm that `search_fields` and `search` override all other search parameters.

    ```bash
    # Overrides search_name_or_sku
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=name&search=blue&search_name_or_sku=scarf"

    # Overrides search and search_sku
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_fields=name&search=blue&search=red&search_sku=TEE"
    ```

7.  **Verify regression:** Confirm that the older search parameters continue to work as expected when the new one is not present.

    *   The following returns the "Casual Blue Shirt" product:
        ```bash
        curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku&search_name_or_sku=blue+shirt"
        ```

    *   The following returns the "Red Cotton Tee" product:
        ```bash
        curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku&search=red"
        ```

    *   The following returns the "Premium Wool Scarf" product:
        ```bash
        curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku&search_sku=SCARF"
        ```

## Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Add support for dynamic multi-field product search via `search_fields` and `search` parameters in the Products REST API endpoint with cross-field matching capability.

</details>

staskus committed:
Add dynamic field searching for products via REST API `search_fields` parameter

Introduces a flexible search mechanism for the v3 products endpoint by adding a `search_fields` parameter.

Previously, searching was limited to predefined parameter combinations like `search_name_or_sku`. This change allows clients to specify exactly which fields to search—`name`, `sku`, or `global_unique_id`—using the `search_fields` parameter in combination with the `search` parameter.

Key features:
- All search tokens must match within the same field (e.g., searching for "blue shirt" will not match a product named "blue" with a SKU of "shirt").
- The `search_fields` parameter takes precedence over all other search-related parameters to ensure predictable results.
- Full backward compatibility is maintained for existing search parameters.

Includes a comprehensive suite of PHPUnit tests to validate the new functionality, including parameter precedence, same-field matching logic, and security against invalid field names.

staskus committed:
Simplify dynamic search clause construction with mapping array

Refactored the build_dynamic_search_clauses method in the products REST API controller to use a mapping array for resolving API field names to database columns, replacing the previous switch statement.

This change makes it easier to extend the search functionality in the future and reduces code repetition.

jorgeatorres commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -311,6 +319,30 @@ protected function prepare_objects_query( $request ) {
+			if ( wc_product_sku_enabled() ) {
+				$allowed_fields[] = 'sku';
+			}
+			$search_fields = array_intersect( $search_fields, $allowed_fields );
```
I don't think this check is necessary. The REST API will take care of validating that `search_fields` contain only the allowed values.

jorgeatorres commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -1799,6 +1871,18 @@ public function get_collection_params() {
+			'type'              => 'array',
+			'items'             => array(
+				'type' => 'string',
+				'enum' => array( 'name', 'sku', 'global_unique_id' ),
```
We can conditionally add `sku` to this enum only when `wc_product_sku_enabled()`.

jorgeatorres reviewed (commented):
Hey @staskus!

This is looking really good and tests as described. I left some feedback but nothing too serious.
I'm wondering whether `search_fields` is the best name for the argument, not because it is incorrect, but because it could give the impression that it can be used to search by _any_ product field which is not the case. Even if that were possible, the search heuristics for these specific name/SKU/GTIN searches are not necessarily what others might expect for a generic search field. Not sure if something like `sku_search_fields` could work (even if `SKU` is not the only field involved).

If we can't find a different arg name, that's probably fine, though. Just curious if you have other ideas.

In any case, I think this implementation is way better than adding new args for every possible combination of fields. Let me know what you think!

staskus committed:
Remove unnecessary manual validation of search_fields parameter

The REST API framework already handles validation of search_fields against
the allowed enum values, making the manual array_intersect validation
redundant. This simplifies the code by removing the duplicate validation
logic and relying on the framework's built-in validation.

staskus committed:
Make SKU field conditionally available in search_fields enum

Only include 'sku' in the search_fields enum when wc_product_sku_enabled()
returns true. This ensures consistency with other SKU-related functionality
and prevents SKU searches when the SKU feature is disabled.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

staskus commented:
@jorgeatorres, thanks for the review! I made improvements based on your comments.

Regarding the naming, it's complicated.

I agree that `search_fields` can sound broad, and its matching logic serves a specific search purpose, although I find it hard to come up with a good alternative. If we put a specific parameter (`sku`, `name`, `global_id`) into this parameter naming, we would come back to the problems of the previous implementation, making `search_fields` inflexible and closed for easy expansion. 

- `search_by` is a possible alternative, although it's similar to `search_fields` in its broadness.
- `search_by_identifier` could communicate that we limit options by what we consider to be an identifier rather than every field.

> Even if that were possible, the search heuristics for these specific name/SKU/GTIN searches are not necessarily what others might expect for a generic search field

Yes, that's an extra level of complexity. It's quite hard to tell whether the search tokens should or shouldn't be split between the SKU/Global Identifier/Name. We assume that you're looking for one of those and not aiming to look for something among multiple fields. Probably `search=` is the one designed for the generic behavior, and `search_fields` could be the more strict one.  Could we cover the clarity with an `API` documentation? 🤔

staskus committed:
Merge branch 'trunk' into woomob-713-add-global_unique_id-to-searched-fields-2

staskus committed:
Linter fixes for class-wc-rest-products-controller.php

staskus commented:
@jorgeatorres Another idea proposed by the team is to keep `search_fields=` but expand it to support the same fields that `search=`  does, which are `name`, `description`, and `short_description`. Then, `search_fields` would support `[name, description, short_description, sku, global_unique_id]`.

 As I understood from the previous comment, `search_fields=name, description, short_description` may not behave the same way as `search=` without passing `search_fields` since it may use a different matching logic, although I haven't checked or tested it yet.

What would be your take on this idea?

staskus committed:
Update Products REST API search_fields with content fields and cross-field matching

- Add description and short_description fields to search_fields parameter
- Implement cross-field matching behavior (tokens match across any specified field)
- Update schema enum to include new content fields
- Add comprehensive test coverage for new functionality
- Maintain WordPress native search compatibility
- Fix code formatting and alignment issues

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

staskus committed:
Merge branch 'trunk' into woomob-713-add-global_unique_id-to-searched-fields-2

staskus committed:
Fix double-escaping of tokens

Remove array_map('esc_sql', ...) calls in prepare_objects_query method that                                                                        were causing double-escaping of search tokens before passing to $wpdb->prepare().                                                                  This was preventing searches with special characters like apostrophes from                                                                        working correctly.

staskus committed:
Merge branch 'trunk' into woomob-713-add-global_unique_id-to-searched-fields-2

jorgeatorres commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -504,32 +532,69 @@ public function add_search_criteria_to_wp_query_join( $join ) {
+				}
+
+				$db_column             = $column_map[ $field ];
+				$field_token_clauses[] = '(' . $db_column . ' LIKE ' . $wpdb->prepare( '%s', $like_search ) . ')';
```
Maybe something along these lines would be a tiny bit easier to grasp?

```suggestion
				$field_token_clauses[] = $wpdb->prepare( "({$db_column} LIKE %s)", $like_search );
```

staskus replied:
Thanks, definitely reads better 👍

jorgeatorres reviewed (approved):
Hey @staskus!

Thanks for making the changes and for all your work on this and for submitting an update to the REST docs as well.
This looks good to me and I'm pre-approving with just one tiny (optional) feedback.

As discussed, this does search the way matching works. Previously searching for, say, `scarf+987` using `search_name_or_sku` wouldn't have returned a match for a product with name "Scarf" and SKU "SKU987" but now it does. As long as this behavior is not problematic for you on the POS side, I'm good with it.

Let me know what you think.

jorgeatorres commented:
> Should I know anything about deployment or merging is enough?

Merging should be enough after the approval.

staskus committed:
Merge branch 'trunk' into woomob-713-add-global_unique_id-to-searched-fields-2

staskus committed:
Improve wpdb::prepare usage in Products Controller search functionality

Refactor SQL query construction in build_dynamic_search_clauses() to use more common WordPress coding patterns. Replace string concatenation approach with wpdb::prepare() for the entire SQL clause.

staskus committed:
Merge branch 'trunk' into woomob-713-add-global_unique_id-to-searched-fields-2

staskus committed:
Add phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared since db_column is whitelisted
